plugins/ltac/g_ltac.mlg

```diff
@@ -409,34 +389,34 @@ END
 
 {
 
-let is_anonymous_abstract = function
-  | TacAbstract (_,None) -> true
-  | TacSolve [TacAbstract (_,None)] -> true
-  | _ -> false
 let rm_abstract = function
-  | TacAbstract (t,_) -> t
-  | TacSolve [TacAbstract (t,_)] -> TacSolve [t]
-  | x -> x
+  | TacAbstract (t,_) -> t, true, false
+  | TacSolve [TacAbstract (t,_)] -> TacSolve [t], true, true
+  | x -> x, false, false
 let is_explicit_terminator = function TacSolve _ -> true | _ -> false
 
 }
 
 VERNAC { tactic_mode } EXTEND VernacSolve STATE proof
-| [ ltac_selector_opt(g) ltac_info_opt(n) tactic(t) ltac_use_default(def) ] =>
+| [ ltac_selector_opt(g) ltac_info_opt(info) tactic(t) ltac_use_default(with_end_tac) ] =>
     { classify_as_proofstep } -> {
     let g = Option.default (Goal_select.get_default_goal_selector ()) g in
-    vernac_solve g n t def
+    let global = match g with Goal_select.SelectAll | Goal_select.SelectList _ -> true | _ -> false in
+    let t = Tacinterp.hide_interp global t None in
+    ComTactic.solve g ~info t ~with_end_tac
   }
-| [ "par" ":" ltac_info_opt(n) tactic(t) ltac_use_default(def) ] =>
+END
+
+VERNAC { tactic_mode } EXTEND VernacSolveParallel STATE proof_ast
+| [ "par" ":" ltac_info_opt(info) tactic(t) ltac_use_default(with_end_tac) ] =>
     {
-      let anon_abstracting_tac = is_anonymous_abstract t in
       let solving_tac = is_explicit_terminator t in
-      let parallel = `Yes (solving_tac,anon_abstracting_tac) in
       let pbr = if solving_tac then Some "par" else None in
-      VtProofStep{ parallel = parallel; proof_block_detection = pbr }
+      VtProofStep{ proof_block_detection = pbr }
     } -> {
-      let t = rm_abstract t in
-      vernac_solve Goal_select.SelectAll n t def
+      let t, abstract, solve = rm_abstract t in
+      let t = Tacinterp.hide_interp true t None in
+      ComTactic.solve_parallel ~info ~ast t ~abstract ~solve ~with_end_tac
```

Comment:
Why don't we expose `Stm.solve_parallel` and call it here instead of going through a hook?

Reply:
Because vernac does not see the STM, it is the other way around.

Reply:
this is ltac not vernac

Reply:
Sorry, I was not clear. Two reasons:
- all commands entry point is in vernac, solve should be no special imo
- at some point I'd like to replace the STM with something else, or even compile coqc without it. I can't really mention the STM from ltac and still do that.

Reply:
I think @SkySkimmer has a point, not sure this is the approach we want to follow here.